File: bionic-js-ios/Bjs/BjsModules.swift
/**
 * Copyright (c) Forge Srl - All Rights Reserved
 * Unauthorized copying, distribution, alteration, transmission or other use of this file is strictly prohibited
 */

import JavaScriptCore

class BjsModules : BjsContext {
    
    let bootstrapFile = "bootstrap.js"
    let bootstrapBundle: BjsBundle
    let appBundle: BjsBundle
    var nativeWrappers: [String : BjsNativeWrapper.Type]
    
    var _nodeJsModule: JSValue?
    var nodeJsModule: JSValue {
        if _nodeJsModule == nil { runBootstrapCode() }
        return _nodeJsModule!
    }
    
    init(_ appBundle: BjsBundle, _ bootstrapBundleName: String = "BjsBootstrap") {
        self.appBundle = appBundle
        bootstrapBundle = BjsBundle(BjsModules.self, bootstrapBundleName)
        nativeWrappers = [String : BjsNativeWrapper.Type]()
    }
    
    func load(_ modulePath: String) -> JSValue {
        
        let jsValue = nodeJsModule.invokeMethod("_load", withArguments: [modulePath, createJsNull() as Any, false])!
        if jsValue.isUndefined {
            logError("cannot load js module \"\(modulePath)\"")
        }
        return jsValue
    }
    
    func addNativeWrappers(_ nativeWrapperClass: BjsNativeWrapper.Type) {
        
        if nativeWrapperClass.name == "" || nativeWrapperClass.wrapperPath == "" {
            fatalError("invalid native wrapper")
        }
        
        if nativeWrappers[nativeWrapperClass.name] != nil {
            fatalError("native wrapper \"\(nativeWrapperClass.name)\" was already added")
        }
        nativeWrappers[nativeWrapperClass.name] = nativeWrapperClass
    }
    
    func removeAllNativeWrappers() {
        nativeWrappers.removeAll()
    }
    
    func clearNodeLoader() {
        _nodeJsModule = nil
    }
    
    private func runBootstrapCode() {
        
        let nativeExport = createNativeExports()
            .exportFunction("executeContent", executeContentFunction())
            .exportFunction("executeFile", executeFileFunction())
            .exportFunction("fileStat", fileStatFunction())
            .exportFunction("readFile", readFileFunction())
            .exportFunction("isInternalModule", isInternalModuleFunction())
            .exportFunction("getInternalModule", getInternalModuleFunction())
        
        guard let bootstrapCode = bootstrapBundle.loadFile(bootstrapFile) else {
            
            logError("cannot find '\(bootstrapFile)' in bundle '\(appBundle.name)'")
            return
        }
        
        // Bootstrap JS file contains a function that takes native dependencies as parameter
        let bootstrapFunction = executeJs("(\(bootstrapCode))", "bundles/\(appBundle.name)/\(bootstrapFile)")
        _nodeJsModule = bootstrapFunction?.call(withArguments: [nativeExport.nativeObj])
    }
    
    private func executeContentFunction() -> @convention(block) (_ : String, _ : String, _ : JSValue, _ : JSValue) -> () {
        
        return { fileContent, filePath, globals, tailCode in
            
            self.executeJsContent(fileContent: fileContent, filePath: filePath, globals: globals, tailCode: tailCode)
        }
    }
    
    private func executeFileFunction() -> @convention(block) (_ : String, _ : JSValue, _ : JSValue) -> () {
        
        return { filePath, globals, tailCode in
            
            guard let fileContent = self.bootstrapBundle.loadFile(filePath) else {
                
                self.setJsExceptionToThrow("js file not found: \(filePath)")
                return
            }
            self.executeJsContent(fileContent: fileContent, filePath: filePath, globals: globals, tailCode: tailCode)
        }
    }
    
    private func executeJsContent (fileContent : String, filePath: String, globals: JSValue, tailCode: JSValue) {
        
        var globalsNames = [String]()
        var globalsValues = [JSValue]()
        let moduleGlobalsLength = globals.objectForKeyedSubscript("length").toNumber().intValue
        for i in 0...moduleGlobalsLength - 1 {
            
            let global = globals.atIndex(i)!
            globalsNames.append(global.atIndex(0).toString())
            globalsValues.append(global.atIndex(1))
        }
        let globalParameters = globalsNames.joined(separator: ",")
        
        let tailCodeString = tailCode.isUndefined ? "" : "\n" + tailCode.toString()
        
        // NodeJS modules loaded during bootstrap are wrapped inside a function and called with arguments provided
        // by JS in bootstrapFile
        
        let functionToCall = self.executeJs("(function(\(globalParameters)){\(fileContent)\(tailCodeString)})", filePath)
        _ = functionToCall?.call(withArguments: globalsValues)
        
        if let internalException = getJsException() {
            setJsExceptionToThrow("error compiling \(filePath): \(internalException.description)")
        }
    }
    
    private func fileStatFunction() -> @convention(block) (_ : String) -> JSValue {
        
        return { path in
            
            guard let filePath = self.appBundle.getFilePath(path) else {
                
                return self.createJsObject(-2) // File not found code ENOENT = 2
            }
            
            var isDirObjC : ObjCBool = false
            let exists = FileManager.default.fileExists(atPath: filePath, isDirectory: &isDirObjC)
            let isDir = isDirObjC.boolValue
            return self.createJsObject(exists ? isDir ? 1 : 0 : -2) // File not found code ENOENT = 2
        }
    }
    
    private func readFileFunction() -> @convention(block) (_ : String) -> JSValue {
        
        return { path in
            
            guard let fileContent = self.appBundle.loadFile(path) else {
                
                self.setJsExceptionToThrow("file not found: \(path)")
                return self.createJsNull()
            }
            return self.createJsObject(fileContent)
        }
    }
    
    private func isInternalModuleFunction() -> @convention(block) (_ : String) -> (Bool) {
        
        return { moduleName in
            return self.nativeWrappers[moduleName] != nil
        }
    }
    
    private func getInternalModuleFunction() -> @convention(block) (_ : String) -> (JSValue) {
        
        return { moduleName in
            return self.nativeWrappers[moduleName]!.bjsGetNativeFunctions(self)!
        }
    }
}

File: bionic-js-java/src/main/java/bionic/js/BjsModules.java
package bionic.js;

import jjbridge.api.runtime.JSReference;
import jjbridge.api.runtime.JSRuntime;
import jjbridge.api.value.JSArray;
import jjbridge.api.value.JSObject;
import jjbridge.api.value.JSString;
import jjbridge.api.value.JSType;
import jjbridge.api.value.strategy.FunctionCallback;

import java.util.HashMap;

class BjsModules extends BjsContext
{
    private static final String bootstrapFile = "bootstrap.js";
    private JSReference nodeJsModule;
    private final BjsBundle bootstrapBundle;
    private final HashMap<String, BjsNativeWrapperTypeInfo<?>> nativeWrappers;
    final BjsBundle appBundle;

    BjsModules(JSRuntime runtime, BjsBundle appBundle)
    {
        this(runtime, appBundle, "BjsBootstrap");
    }

    BjsModules(JSRuntime runtime, BjsBundle appBundle, String bootstrapBundleName)
    {
        super(runtime);
        this.appBundle = appBundle;
        this.bootstrapBundle = new BjsBundle(getClass(), bootstrapBundleName);
        this.nativeWrappers = new HashMap<>();
    }

    JSReference load(String modulePath)
    {
        JSObject<?> nodeJsModule = resolve(getNodeJsModule());
        JSReference load = nodeJsModule.get("_load");
        return callFunction(load, load, newString(modulePath), createJsNull(), newBoolean(false));
    }

    <B extends BjsNativeWrapper<?>> void addNativeWrappers(BjsNativeWrapperTypeInfo<B> nativeWrapperClass)
    {
        if (nativeWrapperClass.name.isEmpty() || nativeWrapperClass.wrapperPath.isEmpty())
        {
            throw new RuntimeException("invalid native wrapper");
        }

        if (nativeWrappers.containsKey(nativeWrapperClass.name))
        {
            throw new RuntimeException("native wrapper " + nativeWrapperClass.name + " was already added");
        }

        nativeWrappers.put(nativeWrapperClass.name, nativeWrapperClass);
    }

    void removeAllNativeWrappers()
    {
        nativeWrappers.clear();
    }

    void clearNodeLoader()
    {
        nodeJsModule = null;
    }

    private JSReference getNodeJsModule()
    {
        if (nodeJsModule == null)
        {
            runBootstrapCode();
        }
        return nodeJsModule;
    }

    private void runBootstrapCode()
    {
        BjsNativeExports nativeExport = createNativeExports()
                .exportFunction("executeContent", executeContentFunction())
                .exportFunction("executeFile", executeFileFunction())
                .exportFunction("fileStat", fileStatFunction())
                .exportFunction("readFile", readFileFunction())
                .exportFunction("isInternalModule", isInternalModuleFunction())
                .exportFunction("getInternalModule", getInternalModuleFunction());

        String bootstrapCode = bootstrapBundle.loadFile(bootstrapFile);
        if (bootstrapCode == null)
        {
            throw new RuntimeException(String.format("Cannot find '%s' in bundle '%s'", bootstrapFile, appBundle.name));
        }

        JSReference bootstrapFunction = executeJs("(" + bootstrapCode + ")",
                String.format("bundle/%s/%s", appBundle.name, bootstrapFile));
        nodeJsModule = callFunction(bootstrapFunction, bootstrapFunction, nativeExport.getExportsObject());
    }

    private FunctionCallback<?> executeContentFunction()
    {
        return jsReferences ->
        {
            String fileContent = ((JSString) resolve(jsReferences[0])).getValue();
            String filePath = ((JSString) resolve(jsReferences[1])).getValue();
            JSReference tailCode = jsReferences.length == 4 ? jsReferences[3] : createJsUndefined();
            executeJsContent(fileContent, filePath, jsReferences[2], tailCode);

            return createJsUndefined();
        };
    }

    private FunctionCallback<?> executeFileFunction()
    {
        return jsReferences ->
        {
            String filePath = ((JSString) resolve(jsReferences[0])).getValue();

            String fileContent = bootstrapBundle.loadFile(filePath);
            if (fileContent == null)
            {
                throw new RuntimeException("JS file not found: " + filePath);
            }

            JSReference tailCode = jsReferences.length == 3 ? jsReferences[2] : createJsUndefined();
            executeJsContent(fileContent, filePath, jsReferences[1], tailCode);

            return createJsUndefined();
        };
    }

    private void executeJsContent(String fileContent, String filePath, JSReference globals, JSReference tailCode)
    {
        JSArray<?> moduleGlobals = resolve(globals);
        int size = moduleGlobals.size();
        String[] globalsNames = new String[size];
        JSReference[] globalsValues = new JSReference[size];

        for (int i = 0; i < size; i++)
        {
            JSArray<?> global = resolve(moduleGlobals.get(i));
            globalsNames[i] = ((JSString) resolve(global.get(0))).getValue();
            globalsValues[i] = global.get(1);
        }

        String globalParameters = String.join(",", globalsNames);
        String tailCodeString = tailCode.getNominalType() == JSType.Undefined
                ? ""
                : "\n" + ((JSString) resolve(tailCode)).getValue();

        String function = String.format("(function(%s){%s%s})", globalParameters, fileContent, tailCodeString);
        JSReference functionToCall = executeJs(function, filePath);
        callFunction(functionToCall, functionToCall, globalsValues);
    }

    private FunctionCallback<?> fileStatFunction()
    {
        return jsReferences ->
        {
            String path = ((JSString) resolve(jsReferences[0])).getValue();
            return newInteger(appBundle.getFileStat(path));
        };
    }

    private FunctionCallback<?> readFileFunction()
    {
        return jsReferences ->
        {
            String path = ((JSString) resolve(jsReferences[0])).getValue();

            String fileContent = appBundle.loadFile(path);
            if (fileContent == null)
            {
                throw new RuntimeException("File not found: " + path);
            }

            return newString(fileContent);
        };
    }

    private FunctionCallback<?> isInternalModuleFunction()
    {
        return jsReferences ->
        {
            String moduleName = ((JSString) resolve(jsReferences[0])).getValue();
            return newBoolean(nativeWrappers.containsKey(moduleName));
        };
    }

    private FunctionCallback<?> getInternalModuleFunction()
    {
        return jsReferences ->
        {
            String moduleName = ((JSString) resolve(jsReferences[0])).getValue();
            return nativeWrappers.get(moduleName).bjsGetNativeFunctions(this);
        };
    }
}
